src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -87,63 +96,91 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+                var arguments = (SeparatedSyntaxList<TArgumentSyntax>)syntaxFacts.GetArgumentsOfInvocationExpression(invocationExpression);
+                return new RegisterFixData<TArgumentSyntax>(arguments, candidates);
+            }
 
-            var arguments = (SeparatedSyntaxList<TArgumentSyntax>)syntaxFacts.GetArgumentsOfInvocationExpression(invocationExpression);
-            var argumentInsertPositionInMethodCandidates = GetArgumentInsertPositionForMethodCandidates(
-                argumentOpt, semanticModel, syntaxFacts, arguments, candidates);
-            RegisterFixForMethodOverloads(context, arguments, argumentInsertPositionInMethodCandidates);
+            return null;
         }
 
-        private async Task HandleObjectCreationExpressionAsync(
-            CodeFixContext context,
-            TObjectCreationExpressionSyntax objectCreation,
-            TArgumentSyntax argumentOpt)
+        private static RegisterFixData<TArgumentSyntax> TryGetObjectCreationFixInfo(
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode node,
+            CancellationToken cancellationToken)
         {
-            var document = context.Document;
-            var cancellationToken = context.CancellationToken;
-            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-
-            // Not supported if this is "new { ... }" (as there are no parameters at all.
-            var typeNode = syntaxFacts.GetObjectCreationType(objectCreation);
-            if (typeNode == null)
+            if (node is TObjectCreationExpressionSyntax objectCreation)
             {
-                return;
-            }
 
-            // If we can't figure out the type being created, or the type isn't in source,
-            // then there's nothing we can do.
-            var type = semanticModel.GetSymbolInfo(typeNode, cancellationToken).GetAnySymbol() as INamedTypeSymbol;
-            if (type == null)
-            {
-                return;
+                // Not supported if this is "new { ... }" (as there are no parameters at all.
+                var typeNode = syntaxFacts.GetObjectCreationType(objectCreation);
+                if (typeNode == null)
+                {
+                    return new RegisterFixData<TArgumentSyntax>();
+                }
+
+                // If we can't figure out the type being created, or the type isn't in source,
+                // then there's nothing we can do.
+                if (!(semanticModel.GetSymbolInfo(typeNode, cancellationToken).GetAnySymbol() is INamedTypeSymbol type))
+                {
+                    return new RegisterFixData<TArgumentSyntax>();
+                }
+
+                if (!type.IsNonImplicitAndFromSource())
+                {
+                    return new RegisterFixData<TArgumentSyntax>();
```

Comment:
instead of allocating, we could just have a singleton empty .Instnace that we return.

Reply:
The lifetime of such a singleton would be unlimited. Is this really better than creating a new short lived instance in these rare cases?

Reply:
any reason not just returning null? if you don't want to ever check null and naturally fail due to data is empty, you can change RegisterFixData to struct? seems, it will be fine being struct?

having static Empty field for Null object design pattern is roslyn pattern though.

Reply:
no big deal though. fine with as it is as well.

Reply:
Returning null means "let's try the next parent node", while returning the empty data object means "here is nothing I can do. You can give up (actually ending the fixer without registering a code action)".